File: gfg/medium/largest-subarray-with-0-sum/solution.py
class Solution:
    def maxLength(self, arr):
        prefix_sum = 0
        max_len = 0
        first_index = {}

        for i, num in enumerate(arr):
            prefix_sum += num

            # Subarray from index 0 to i has sum 0
            if prefix_sum == 0:
                max_len = i + 1

            # If prefix sum was seen before, the middle part has sum 0
            if prefix_sum in first_index:
                max_len = max(max_len, i - first_index[prefix_sum])
            else:
                first_index[prefix_sum] = i

        return max_len

File: gfg/medium/largest-subarray-with-0-sum/solution.cpp
class Solution {
  public:
    int maxLength(vector<int>& arr) {
        unordered_map<int, int> firstIndex;
        
        int sum = 0;
        int maxLen = 0;
        
        for (int i = 0; i < arr.size(); i++) {
            sum += arr[i];
            
            // If prefix sum is 0, subarray from 0 to i has sum 0
            if (sum == 0) {
                maxLen = i + 1;
            }
            
            // If this sum was seen before, the elements between have sum 0
            if (firstIndex.find(sum) != firstIndex.end()) {
                maxLen = max(maxLen, i - firstIndex[sum]);
            } else {
                firstIndex[sum] = i;
            }
        }
        
        return maxLen;
    }
};